CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,283 @@
+
+    int n = pfcandidates->size();
+    std::unique_ptr<std::vector<reco::PFCandidate>> copy(new std::vector<reco::PFCandidate>());
+    std::unique_ptr<std::vector<reco::PFCandidate>> discarded(new std::vector<reco::PFCandidate>());
+    copy->reserve(n);
+    std::vector<int> oldToNew(n), newToOld, badToOld; 
+    newToOld.reserve(n);
+
+    //std::cout << "NEW EV:" << std::endl;
+
+    //loop over PFCandidates
+    int i = -1;
+    for (const reco::PFCandidate &pf : *pfcandidates) 
+      {
+	++i;
+
+	//deal with HE
+	if( pf.particleId() == 5 && 
+	    fabs(pf.eta()) > 1.4  && fabs(pf.eta()) < 3.)
+	  {
+	    bool toKill = false;
+	    for(auto badIt: badChHE_)
+	      if( reco::deltaR2(pf.eta(), pf.phi(), std::get<0>(badIt), std::get<1>(badIt)) < 0.07 )
+		toKill = true;
+	    
+	    
+	    if(toKill)
+	      {
+		discarded->push_back(pf);
+		oldToNew[i] = (-discarded->size());
+		badToOld.push_back(i);
+		continue;
+	      }
+	    else
+	      {
+		copy->push_back(pf);
+		oldToNew[i] = (copy->size());
+		newToOld.push_back(i);
+	      }
+	  }
+	//deal with HF
+	else if(fabs(pf.eta()) > 3.)
+	  {
+	    math::XYZPointF ecalPoint = pf.positionAtECALEntrance();
+	    GlobalPoint ecalGPoint(ecalPoint.X(),ecalPoint.Y(),ecalPoint.Z());
+	    HcalDetId closestDetId(hgeom->getClosestCell(ecalGPoint));
+	    
+	    if(closestDetId.subdet() == 4)
+	      {
+		HcalDetId hDetId(closestDetId.subdet(),closestDetId.ieta(),closestDetId.iphi(),1); //depth1
+		
+		//raw*calEnergy() is the same as *calEnergy() - no corrections are done for HF
+		float longE  = pf.rawEcalEnergy() + pf.rawHcalEnergy()/2.;  //depth1
+		float shortE = pf.rawHcalEnergy()/2.;                       //depth2
+		
+		float ecalEnergy = pf.rawEcalEnergy();
+		float hcalEnergy = pf.rawHcalEnergy();
+		float totEnergy = ecalEnergy + hcalEnergy;
+		
+		bool toKill = false;
+			
+		for(auto badIt: badChHF_)
+		  {
+		    if ( hDetId.ieta() == std::get<0>(badIt) &&
+			 hDetId.iphi() == std::get<1>(badIt) )
+		      {
+			//std::cout << "==> orig en (tot,H,E): " << pf.energy() << " " << pf.rawHcalEnergy() << " " << pf.rawEcalEnergy() << std::endl;
+			if(std::get<2>(badIt) == 1) //depth1
+			  {
+			    longE *= std::get<3>(badIt);
+			    ecalEnergy = longE - shortE;
+			    totEnergy = ecalEnergy + hcalEnergy;
+			  }
+			else //depth2
+			  {
+			    shortE *= std::get<3>(badIt);
+			    hcalEnergy = 2*shortE;
+			    if(ecalEnergy > 0)
+			      ecalEnergy = longE - shortE;
+			    totEnergy = ecalEnergy + hcalEnergy;
+			  }
+			//kill candidate if goes below thr
+			if((pf.pdgId()==1 && shortE < 1.4) || 
+			   (pf.pdgId()==2 && longE < 1.4))
+			  toKill = true;
+
+			//std::cout << "====> ieta,iphi,depth: " <<std::get<0>(badIt) << " " << std::get<1>(badIt) << " " << std::get<2>(badIt) << " corr: " << std::get<3>(badIt) << std::endl;
+			//std::cout << "====> recal en (tot,H,E): " << totEnergy << " " << hcalEnergy << " " << ecalEnergy << std::endl;
```

Comment:
it's more practical to define a debug flag and have these printouts uncommented  (even better to use edm::LogPrint or edm::LogVerbatim so that this also works in multithreaded calls)